File: src/Sample/Plumix.Sample/Demos/Material/SwitchListTileDemoPage.cs
using System;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source (reference): dart_sample/lib/demos/material/switch_list_tile_demo_page.dart
public sealed class SwitchListTileDemoPage : StatefulWidget
{
    public override State CreateState() => new SwitchListTileDemoPageState();
}

internal sealed class SwitchListTileDemoPageState : State
{
    private bool _switchValue = true;
    private bool _enabled = true;
    private bool _adaptive;
    private bool _compact;
    private ListTileControlAffinity _affinity = ListTileControlAffinity.Trailing;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("SwitchListTile", fontSize: 20, color: Colors.Black),
                new Text(
                    "Whole-tile interaction, affinity, density/alignment, selected styling, "
                    + "disabled state, and the adaptive branch (which paints the Cupertino switch).",
                    fontSize: 14,
                    color: Color.Parse("#8A000000")),
                new Row(
                    spacing: 8,
                    children:
                    [
                        ListTileControlDemos.ControlButton(
                            _enabled ? "Enabled" : "Disabled",
                            () => SetState(() => _enabled = !_enabled),
                            104,
                            Color.Parse("#FFE9F0FF")),
                        ListTileControlDemos.ControlButton(
                            _affinity == ListTileControlAffinity.Leading ? "Leading" : "Trailing",
                            () => SetState(() => _affinity = _affinity == ListTileControlAffinity.Leading
                                ? ListTileControlAffinity.Trailing
                                : ListTileControlAffinity.Leading),
                            104,
                            Color.Parse("#FFE9F7EF")),
                        ListTileControlDemos.ControlButton(
                            _adaptive ? "Adaptive" : "Material",
                            () => SetState(() => _adaptive = !_adaptive),
                            104,
                            Color.Parse("#FFF8EFE2")),
                        ListTileControlDemos.ControlButton(
                            _compact ? "Compact / top" : "Standard / center",
                            () => SetState(() => _compact = !_compact),
                            144,
                            Color.Parse("#FFF3E5F5")),
                    ]),
                new Text(
                    $"switch={ListTileControlDemos.Lower(_switchValue)}, "
                    + $"affinity={_affinity.ToString().ToLowerInvariant()}, "
                    + $"adaptive={ListTileControlDemos.Lower(_adaptive)}",
                    fontSize: 12,
                    color: Color.Parse("#FF607D8B")),
                new Expanded(
                    child: new Container(
                        color: Color.Parse("#FFF7F9FC"),
                        child: new ListTileTheme(
                            data: new ListTileThemeData(
                                ControlAffinity: _affinity,
                                VisualDensity: _compact ? VisualDensity.Compact : VisualDensity.Standard,
                                TitleAlignment: _compact
                                    ? ListTileTitleAlignment.Top
                                    : ListTileTitleAlignment.Center),
                            child: new Column(
                                crossAxisAlignment: CrossAxisAlignment.Stretch,
                                children:
                                [
                                    BuildSwitchTile(),
                                ])))),
            ]);
    }

    private Widget BuildSwitchTile()
    {
        Action<bool>? onChanged = _enabled
            ? value => SetState(() => _switchValue = value)
            : null;
        return _adaptive
            ? SwitchListTile.Adaptive(
                value: _switchValue,
                onChanged: onChanged,
                title: new Text("Background sync"),
                subtitle: new Text("The embedded switch remains draggable."),
                secondary: new Icon(Icons.Menu),
                selected: _switchValue)
            : new SwitchListTile(
                value: _switchValue,
                onChanged: onChanged,
                title: new Text("Background sync"),
                subtitle: new Text("The embedded switch remains draggable."),
                secondary: new Icon(Icons.Menu),
                selected: _switchValue,
                selectedTileColor: Color.Parse("#FFE8DEF8"));
    }
}

File: dart_sample/lib/demos/material/switch_list_tile_demo_page.dart
import 'package:material_ui/material_ui.dart';

class SwitchListTileDemoPage extends StatefulWidget {
  const SwitchListTileDemoPage({super.key});

  @override
  State<SwitchListTileDemoPage> createState() => _SwitchListTileDemoPageState();
}

class _SwitchListTileDemoPageState extends State<SwitchListTileDemoPage> {
  bool _switchValue = true;
  bool _enabled = true;
  bool _adaptive = false;
  bool _compact = false;
  ListTileControlAffinity _affinity = ListTileControlAffinity.trailing;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'SwitchListTile',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Whole-tile interaction, affinity, density/alignment, selected styling, disabled state, and the adaptive branch (which paints the Cupertino switch).',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildControlButton(
              _enabled ? 'Enabled' : 'Disabled',
              () => setState(() => _enabled = !_enabled),
              104,
              const Color(0xFFE9F0FF),
            ),
            _buildControlButton(
              _affinity == ListTileControlAffinity.leading
                  ? 'Leading'
                  : 'Trailing',
              () => setState(() {
                _affinity = _affinity == ListTileControlAffinity.leading
                    ? ListTileControlAffinity.trailing
                    : ListTileControlAffinity.leading;
              }),
              104,
              const Color(0xFFE9F7EF),
            ),
            _buildControlButton(
              _adaptive ? 'Adaptive' : 'Material',
              () => setState(() => _adaptive = !_adaptive),
              104,
              const Color(0xFFF8EFE2),
            ),
            _buildControlButton(
              _compact ? 'Compact / top' : 'Standard / center',
              () => setState(() => _compact = !_compact),
              144,
              const Color(0xFFF3E5F5),
            ),
          ],
        ),
        Text(
          'switch=$_switchValue, affinity=${_affinity.name}, adaptive=$_adaptive',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Expanded(
          child: ColoredBox(
            color: const Color(0xFFF7F9FC),
            child: ListTileTheme(
              data: ListTileThemeData(
                controlAffinity: _affinity,
                visualDensity: _compact
                    ? VisualDensity.compact
                    : VisualDensity.standard,
                titleAlignment: _compact
                    ? ListTileTitleAlignment.top
                    : ListTileTitleAlignment.center,
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: <Widget>[_buildSwitchTile()],
              ),
            ),
          ),
        ),
      ],
    );
  }

  Widget _buildSwitchTile() {
    final ValueChanged<bool>? onChanged = _enabled
        ? (bool value) => setState(() => _switchValue = value)
        : null;
    if (_adaptive) {
      return SwitchListTile.adaptive(
        value: _switchValue,
        onChanged: onChanged,
        title: const Text('Background sync'),
        subtitle: const Text('The embedded switch remains draggable.'),
        secondary: const Icon(Icons.menu),
        selected: _switchValue,
      );
    }
    return SwitchListTile(
      value: _switchValue,
      onChanged: onChanged,
      title: const Text('Background sync'),
      subtitle: const Text('The embedded switch remains draggable.'),
      secondary: const Icon(Icons.menu),
      selected: _switchValue,
      selectedTileColor: const Color(0xFFE8DEF8),
    );
  }

  Widget _buildControlButton(
    String label,
    VoidCallback onPressed,
    double width,
    Color background,
  ) {
    return SizedBox(
      width: width,
      child: TextButton(
        onPressed: onPressed,
        style: TextButton.styleFrom(
          backgroundColor: background,
          foregroundColor: Colors.black,
          minimumSize: const Size(0, 36),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }
}
